Use interconnect handler instead of federation

Martin-Idel opened the pull request:
Since this is the API that should (for now) stay, we should port the example over to use this API.

Martin-Idel committed:
Use interconnect handler instead of federation

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Rename cosmicconnector to (Simple)InterconnectProvider

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Implement example functions for transceivers

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add aip linter

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Try to compile linter directly

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
First pass fixing linter issues

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add go_path target to help IDEs

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Start implementation of example provider

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Remove unsigned types as they are not available always

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Rename static fields as 'static' is reserved in many languages

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add field annotations everywhere

We assume that filters are optional and
all other fields are required for now.

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add missing methods where it seems to make sense

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add pagination information according to documentation

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add breadcrumbs when api-linter-use in Bazel might get easier

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Fixup: Rename comment in Bearer

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add missing references to targets and transceivers

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Revert "Add pagination information according to documentation"

This reverts commit 82d19509206c7ecccef1e36b881ba8fea804dccd.

Martin-Idel committed:
Better comment for leaving pagination as is

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add HTTP information to API

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add some documentation comments

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add missing connection between bearer and ac

In addition, add some more comments to better explain
the corresponding methods in the interconnection service.

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Improve handler example

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Move go_path to example to cover more packages

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

electronjoe commented on MODULE.bazel:
```diff
@@ -78,7 +78,9 @@ go_deps.module(
     version = "v1.68.0",
 )
-# Dependencies of api-lint
+# ---------------------------
```
Excellent, thank you!

electronjoe commented on outernet/federation/interconnect/v1alpha/interconnect.proto:
```diff
@@ -152,7 +154,13 @@ service InterconnectService {
   }
 
-  // Creates a bearer.
+  // Creates a bearer. A bearer defines the physical connection necessary to create attachment circuits (data links)
```
Thanks for the improved documentation!

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
+
 func (p *PrototypeHandler) ListBearers(context.Context, *pb.ListBearersRequest) (*pb.ListBearersResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method ListBearers not implemented")
+	// TODO: Handle filter
```
Here and elsewhere Filter TODO exist.  For now could you check for a non-empty Filter specification and return an error (UNIMPLEMENTED?)

Martin-Idel replied:
Sounds good, will do.

electronjoe commented on examples/golang/simpleinterconnectprovider/BUILD.bazel:
```diff
@@ -35,3 +35,11 @@ go_library(
     ],
 )
+
+go_path(
```
Can you provide a bit more context on what this is doing for us?

Martin-Idel replied:
What this does for me is it allows for much improved IDE support (I can add a comment).

What happens? This creates a "go path" in the bazel-bin folder, which can be passed to the IDE to resolve dependencies. So, for instance, using 
```
{
    "go.gopath": "<path-to-repo>//bazel-bin/examples/golang/simpleinterconnectprovider/gopath"
}
```
inside `.vscode/settings` and I now get correct dependency resolution (including source code) for all imports referenced in the handler target and all its dependencies. It might even be possible to get some more linting going at least locally, but I haven't looked into it, yet.

There are two things I couldn't get going:
- I would've like to just use the main target, but there is a conflict in dependency resolution which I don't understand, yet.
- Imports of test-only dependencies don't work yet, because this target cannot depend on test targets.

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
-	return nil, status.Errorf(codes.Unimplemented, "method GetTarget not implemented")
+
+func (p *PrototypeHandler) GetTarget(_ context.Context, targetRequest *pb.GetTargetRequest) (*pb.Target, error) {
+	if p.target[targetRequest.Name] != nil {
```
I think this condition is inverted? Probably means we have a test coverage hole too?

Martin-Idel replied:
Correct, yes. I just noticed I missed tests for the target functions completely - that's also the other whole with ListTargets... Added all missing tests I could find.

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
 func (p *PrototypeHandler) ListTargets(context.Context, *pb.ListTargetsRequest) (*pb.ListTargetsResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method ListTargets not implemented")
+	targets := make([]*pb.Target, 0, len(p.target))
+	for _, target := range targets {
```
Should this be iterating over p.target? Maybe a test coverage hole.

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
+	defer p.mu.Unlock()
+
+	attachmentCircuitName := fmt.Sprintf("attachmentCircuits/%s", ac.AttachmentCircuitId)
+	if p.bearers[attachmentCircuitName] != nil {
```
Should this be p.attachmentCircuits[attachmentCircuitName]? Perhaps a test coverage hole.

Martin-Idel replied:
Good catch, that's a test I forgot but have for the other parts...

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -55,7 +76,6 @@ func (p *PrototypeHandler) ListCompatibleTransceiverTypes(context.Context, *pb.L
 }
 
 func (p *PrototypeHandler) GetTransceiver(ctx context.Context, trans *pb.GetTransceiverRequest) (*pb.Transceiver, error) {
```
Here and elsewhere.

I think that several of your read-only operations (GetX, ListX) read shared maps (p.transceivers, p.bearers, etc.) without acquiring the mutex. It's entirely possible that my existing example has this same oversight =D Although this is a prototype, and you do lock on writes (create/update/delete), calling these read methods concurrently with a write can lead to data races.

Recommended Fix: Use the same p.mu.Lock()/defer p.mu.Unlock() (or an RLock) pattern for each read method, or transition to a sync.RWMutex and do RLock for gets/lists. For example:

```go
func (p *PrototypeHandler) GetTransceiver(ctx context.Context, trans *pb.GetTransceiverRequest) (*pb.Transceiver, error) {
    p.mu.Lock()
    defer p.mu.Unlock()

    t := p.transceivers[trans.Name]
    if t == nil {
        return nil, status.Errorf(codes.NotFound, "transceiver not found")
    }
    return t, nil
}
```
Doing so ensures that all concurrent reads/writes on the maps remain safe.

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
+				return true
+			}
+
+			if knownBearer.Interval.EndTime.AsTime().After(bearer.Interval.StartTime.AsTime()) && bearer.Interval.StartTime.AsTime().After(knownBearer.Interval.StartTime.AsTime()) ||
```
I believe there may be a clarity issue on Interval use for bearer's and ContactWindows - is it ever documented whether they are closed intervals [start, stop] or half-open [start, stop) etc?  As it impacts this logic and generally I think we probably want half-open so that you can chain them together and they don't overlap?  Maybe an issue for the Proto comments (unless Intervals have already defined this as Google defines them? Maybe they have).

Martin-Idel replied:
Correct. It is half open [start,stop) so I should be more careful here. I'll fix that and I'll add a comment somewhere (everywhere?) in the API so that people don't have to look it up.

Martin-Idel replied:
Although, I THINK my logic is fine. I added a test for time adjacent bearers and that works correctly, so aside from comments we should hopefully be fine. I can't think of a different edge case right now.

electronjoe commented on examples/golang/simpleinterconnectprovider/handler/handler.go:
```diff
@@ -99,39 +168,227 @@ func (p *PrototypeHandler) DeleteTransceiver(_ context.Context, trans *pb.Delete
+			continue
+		}
+
+		if contactWindow.MinRxCenterFrequencyHz >= bearer.RxCenterFrequencyHz ||
```
nit, I believe this logic as declared doesn't allow for equality (e.g. bearer's RxCenterFrequencyHz cannot be equal to contactWindow.MinRxCenterFrequencyHz?).  That seems like a minor typo (I would think it's OK if they are equal)?

Martin-Idel replied:
makes sense. I changed it and added a test for one of the cases.

electronjoe reviewed (approved):
LGTM % comments

Martin-Idel committed:
Fixes after review

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add comment about how intervals work

Intervals are defined to be half-open, i.e.
closed at the beginning and open at the end.
That allows to specify intervals that touch and
thereby don't overlap.

We add a comment that we adhere to that rule
so that people do not have to look it up.

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Update README for examples with new API

Signed-off-by: Martin Idel <martin.idel@tngtech.com>
